### ISSUE TITLE
roi issue with inpaint opposed & crop 

### ISSUE THREAD

**Issue by MStraeten** [darktable-org/darktable#12692]
when using inpaint opposed (as well as segmentation based highlight reconstruction) in combination with crop the right side of the image turns black
[mire1.cr2.xmp.txt](https://github.com/darktable-org/darktable/files/9845109/mire1.cr2.xmp.txt)

the behaviour is dependent on the crop - playing around with more or less crop in hight or width results in proper crops but also in misbehaviour. Quite reproducible bis 16:9 ratio if the crop area has some space to the top ...

**To Reproduce**
1. open mire1.cr2 from tests/images
2. apply inpaint opposed highlight reconstruction
3. apply crop (e.g. 16:9) center aligned

**Screenshots**
![image](https://user-images.githubusercontent.com/39386816/197358537-be4a1271-7138-4b74-818f-0e0c20add5a8.png)

**Which commit introduced the error**
```
1cb4f1afe3a7125955b13a9457b7335a97b5bf9f is the first bad commit
commit 1cb4f1afe3a7125955b13a9457b7335a97b5bf9f
Author: hanno@schwalm-bremen.de <hanno@schwalm-bremen.de>
Date:   Fri Oct 21 19:35:54 2022 +0200

    Proper fix for highlights reconstruction ROI problems

    As reported in #12667 there were crashes in opposed or segmentation based code.

    There has been a previous pr #12670, that one has become a mess but is a
    valuable resource to understand the problems.

    In short; the mentioned algorithms require all image data for a stable
    calculation of chromacity correction (opposed) and the segmentation
    analysis and recovery code.

    Thus we have to change the roi_in to "take it all".
    But - and that has been the problem of current code - the roi_out can be
    extended (for example by demosaic rounding for xtrans sensors).
    This results in a situation where the roi_out area is not fully inside roi_in.

    This can't be properly fixed in the modify code, we have to check for the
    proper output range in the algorithms.

    Fixes #12667

    Replaces #12670

 src/iop/highlights.c    |  19 ++++-----
 src/iop/hlrecovery_v2.c |  58 +++++++++++++++-----------
 src/iop/opposed.c       | 107 +++++++++++++++++++++++++++---------------------
 3 files changed, 101 insertions(+), 83 deletions(-)
```
**Platform**
* darktable version : darktable 4.1.0+614~g832272359
* OS                : OSx

**Comment by da-phil** [comment]
I can confirm the issue.

**Comment by jenshannoschwalm** [comment]
Thanks for reporting :-(((

Can confirm too so more work ...

**Comment by jenshannoschwalm** [comment]
@MStraeten that was a stupid yet very simple one on my side.

**Comment by da-phil** [comment]
@jenshannoschwalm thanks for your fast fix, it works for me ;)

**Comment by MStraeten** [comment]
fixed with #12689 thanks for the fix
